File: testscripts/modbus/fims_decode.py
import sys
import json
from datetime import datetime

def get_timestamp_offset(first_timestamp, current_timestamp):
    first_ts_datetime = datetime.strptime(first_timestamp, "%Y-%m-%d %H:%M:%S.%f")
    current_ts_datetime = datetime.strptime(current_timestamp, "%Y-%m-%d %H:%M:%S.%f")
    time_delta = current_ts_datetime - first_ts_datetime
    return time_delta.total_seconds()

def parse_uri(uri):
    # Assuming the URI segments are separated by '/'
    segments = uri.split('/')
    if len(segments) >= 3:
        last_segment = segments[-1]
        new_uri = '/'.join(segments[:-1])
        return last_segment, new_uri
    return None, uri

def decode_input_file(input_file):
    with open(input_file, 'r') as f:
        data = f.read()

    entries = data.strip().split("\n\n")
    json_objects = {}

    first_timestamp = None

    for entry in entries:
        json_object = {}
        lines = entry.strip().split("\n")
        for line in lines:
            key, value = map(str.strip, line.split(":", 1))
            if key == "Body":
                valueOK = True
                try:
                    value = json.loads(value)
                except json.JSONDecodeError:
                    valueOK = False

                if not isinstance(value, dict):
                    valueOK = False

                if not valueOK:
                    last_segment, new_uri = parse_uri(json_object.get("Uri"))
                    if last_segment:
                        value = {last_segment: value}
                        json_object["Uri"] = new_uri

            json_object[key] = value
            if key == "Timestamp":
                if not first_timestamp:
                    first_timestamp = value

                timestamp_offset = get_timestamp_offset(first_timestamp, value)
                json_objects[timestamp_offset] = json_object

    return json_objects

def main(input_file, output_file):
    json_objects = decode_input_file(input_file)

    with open(output_file, 'w') as f:
        json.dump(json_objects, f, indent=2)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python fims_decode.py input_file output_file")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    main(input_file, output_file)

File: testscripts/modbus/fims_decode.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <ctime>
#include <iomanip>
#include <nlohmann/json.hpp> // Include header for the json library (https://github.com/nlohmann/json)

double getTimestampOffset(const std::string& firstTimestamp, const std::string& currentTimestamp) {
    std::tm firstTs = {};
    std::tm currentTs = {};
    std::istringstream firstStream(firstTimestamp);
    std::istringstream currentStream(currentTimestamp);

    firstStream >> std::get_time(&firstTs, "%Y-%m-%d %H:%M:%S");
    currentStream >> std::get_time(&currentTs, "%Y-%m-%d %H:%M:%S");

    std::time_t firstTime = mktime(&firstTs);
    std::time_t currentTime = mktime(&currentTs);
    return std::difftime(currentTime, firstTime);
}

std::pair<std::string, std::string> parseUri(const std::string& uri) {
    std::stringstream ss(uri);
    std::string segment;
    std::vector<std::string> segments;
    while (std::getline(ss, segment, '/')) {
        segments.push_back(segment);
    }

    if (segments.size() >= 3) {
        std::string lastSegment = segments.back();
        segments.pop_back();
        std::string newUri = "/" + std::accumulate(segments.begin(), segments.end(), std::string(),
            [](const std::string& a, const std::string& b) {
                return a + "/" + b;
            });
        return {lastSegment, newUri};
    }
    return {"", uri};
}

nlohmann::json decodeInputFile(const std::string& inputFile) {
    std::ifstream file(inputFile);
    std::string line;
    std::vector<std::string> entries;
    nlohmann::json jsonObjects;

    std::string entry;
    while (std::getline(file, line)) {
        if (line.empty() && !entry.empty()) {
            entries.push_back(entry);
            entry.clear();
        } else {
            entry += line + "\n";
        }
    }
    if (!entry.empty()) {
        entries.push_back(entry);
    }

    std::string firstTimestamp;

    for (const auto& entry : entries) {
        std::stringstream ss(entry);
        nlohmann::json jsonObject;
        while (std::getline(ss, line)) {
            size_t pos = line.find(":");
            std::string key = line.substr(0, pos);
            std::string value = line.substr(pos + 1);
            
            // Trim whitespaces (from the beginning and the end)
            key.erase(0, key.find_first_not_of(" \n\r\t"));
            key.erase(key.find_last_not_of(" \n\r\t") + 1);
            value.erase(0, value.find_first_not_of(" \n\r\t"));
            value.erase(value.find_last_not_of(" \n\r\t") + 1);

            if (key == "Body") {
                try {
                    jsonObject[key] = nlohmann::json::parse(value);
                } catch (nlohmann::json::parse_error& e) {
                    auto [lastSegment, newUri] = parseUri(jsonObject["Uri"]);
                    if (!lastSegment.empty()) {
                        jsonObject["Uri"] = newUri;
                        jsonObject[key] = nlohmann::json{{lastSegment, value}};
                    }
                }
            } else if (key == "Timestamp") {
                if (firstTimestamp.empty()) {
                    firstTimestamp = value;
                }
                double timestampOffset = getTimestampOffset(firstTimestamp, value);
                jsonObjects[std::to_string(timestampOffset)] = jsonObject;
            } else {
                jsonObject[key] = value;
            }
        }
    }
    return jsonObjects;
}

int main(int argc, char** argv) {
    if (argc != 3) {
        std::cout << "Usage: ./fims_decode input_file output_file" << std::endl;
        return 1;
    }

    std::string inputFile = argv[1];
    std::string outputFile = argv[2];

    nlohmann::json jsonObjects = decodeInputFile(inputFile);

    std::ofstream outFile(outputFile);
    outFile << std::setw(2) << jsonObjects << std::endl;
    return 0;
}
